refactor(stock): merge stock receipts

jniles opened the pull request:
Merges most stock receipt into a single URL to allow the client to uniformly render the stock receipt with a single API call.

Closes #4073.

---

This PR now contains a breaking change to stock_movements that should probably result in a new release.

The way we calculate stock_movement references has changed.  Instead of a new `reference` for each line of the stock movement, references now behave like `trans_id` or cash/voucher/invoice references, being assigned one per stock_movement.  This allows us to map `document_uuid` one-to-one with the `stock_movement` reference.

The PRIMARY KEY is now the `uuid`, and `document_uuid` also has an INDEX added to allow quick lookups.

Finally, the calculation of `reference` has been moved to a trigger and the associated stored procedures have been removed.  We've also modified the `zRecomputeDocumentMap()` procedure to re-generate the document mappings from the `stock_movement` table.

jniles committed:
refactor(stock): merge stock receipts

Merges most stock receipt into a single URL to allow the client to
uniformly render the stock receipt with a single API call.

Closes #4073.

jniles committed:
fix(stock): add header to stock movement menu

Adds a header to the dropdown on the stock movement registry to better
tie each row with its movement reference.

mbayopanda commented on server/controllers/stock/reports/index.js:
```diff
@@ -26,24 +30,106 @@ const stockValue = require('./stock/value');
+      renderer = stockAdjustmentReceipt;
+      break;
+
+    case Stock.flux.FROM_INTEGRATION:
```
There is no `break;` here

jniles replied:
I assumed that FROM_INTEGRATION and TO_INTEGRATION used the same receipt, so I combined them.  If they don't, what receipt does stock exit via integration use?

mbayopanda reviewed (commented):
@jniles i cannot see these pdf receipts : 
- stock exit to service receipt for default tests data
- stock exit to loss receipt for default tests data
- stock adjustment receipt




![image](https://user-images.githubusercontent.com/5445251/72264296-caf00a80-361a-11ea-8a3f-8293079a9fef.png)

jniles committed:
fix: properly reference stock adjustment

Adds in the receipt for stock adjustments to the receipt rendering
matrix.

jniles committed:
fix: compute movement references by trigger

Adds the ability to compute stock movement references by trigger, not by
calling a stored procedure.  The change reduces the amount of code and
increases the assurance that document_maps will exist for each stock
movement.

jniles committed:
fix(stock): standard movement reference creation

BREAKING CHANGE

Changes the way stock movements have their references created by using
a trigger instead of a stored procedure.  Also update
zRecomputeDocumentMap() to take into account stock movements and
generate the document mappings when requested for these movements.

I've also fundamentally changed the PRIMARY KEY of the stock_movement
table from an integer (!) to the UUID of the stock movement.  I've also
added an INDEX on the document_uuid, since we'll probably be looking up
the document_uuid the most.

mbayopanda commented on server/controllers/stock/reports/index.js:
```diff
@@ -26,24 +30,107 @@ const stockValue = require('./stock/value');
+ * Figures out the type of stock receipt from the document uuid.  This allows
+ * a uniform API for rendering all stock receipts.
+ */
+async function determineReceiptType(uuid, isDepotTransferExit = -1) {
```
👍 that a good idea to figure out the type of stock receipt just from the document uuid.

mbayopanda reviewed (approved):
LGTM 👍 

Great Job Jon

bors r+
